### PR TITLE
UML-3750 added exception for null DOB

### PR DESCRIPTION
**Pull request by allenannom**
# Purpose

_Added an exception to be thrown when the DOB is null. Then implemented a try/catch with a warning log_

Fixes UML-3750

## Approach

appropriately outputs the known error and logs _

## Checklist

* [ ] I have performed a self-review of my own code
* [ ] I have added relevant logging with appropriate levels to my code
* [ ] New event_codes have been documented on the [wiki page](https://opgtransform.atlassian.net/wiki/spaces/LSML2/pages/3277881441/Understanding+the+event+logs)
* [ ] I have updated documentation (Confluence/GitHub wiki/tech debt doc) where relevant
* [ ] I have added tests to prove my work
* [ ] I have added welsh translation tags and updated translation files
* [ ] I have run an accessibility tool on any pages I have made changes to and fixed any issues found
* [ ] I have notified the Interaction Designer of any content changes so that appropriate screenshots/flow diagram changes can be made
* [ ] The product team have tested these changes

### CODE REVIEW

**Review by SeemaMenon — APPROVED** [review]

File: service-api/app/src/App/src/Service/Lpa/SiriusPerson.php [hunk 1]
```diff
@@ -80,8 +81,15 @@ public function getPostcode(): string
         return (string)$this->person['addresses'][0]['postcode'];
     }
 
+    /**
+     * @throws Exception
+     */
     public function getDob(): DateTimeInterface
     {
+        if (is_null($this->person['dob'])) {
+            throw new Exception('Actor DOB is not set');
```
Comment (cooperaj) [inline service-api/app/src/App/src/Service/Lpa/SiriusPerson.php:90]: I think a custom exception should be created/thrown for this. Something like `ActorDateOfBirthNotSetException`

File: service-api/app/src/App/src/Service/Lpa/FindActorInLpa.php [hunk 2]
```diff
@@ -137,7 +138,22 @@ private function checkForActorMatch(ActorMatchingInterface $actor, array $matchD
 
         $match = self::MATCH;
 
-        $match = $actor->getDob()->format('Y-m-d') !== $matchData['dob']
+        try {
```
Comment (cooperaj) [inline service-api/app/src/App/src/Service/Lpa/FindActorInLpa.php:141]: Rather than have the exception just wrap the problematic call my preference is to have them wrap as much code as is relevant for the operation.

In this case I'd toy with putting the entirety of the rest of the function in it or maybe lines 156 - 167, or for the smallest amount 156-158. Basically avoid having to create new variables to make the exception trapping happen (in this case ActorDob)

**Review by cooperaj — COMMENTED** [review]
Seems to be missing the test case in the patch I sent that covers an LPA continuing to match on good actors even if a null dob actor is included.

File: service-api/app/src/App/src/Service/Lpa/FindActorInLpa.php [hunk 3]
```diff
@@ -137,7 +138,22 @@ private function checkForActorMatch(ActorMatchingInterface $actor, array $matchD
...
+            $actorDob = $actor->getDob();
+        } catch (Exception $e) {
+            $this->logger->warning(
+                'Actor DOB is null',
+                [
+                    'actor_id'   => $actor->getUid(),
+                    'actor_data' => $actorData,
```
Comment (mattmachell) [inline service-api/app/src/App/src/Service/Lpa/FindActorInLpa.php:148]: BY my reading this will log a bunch of PII, so I'd remove it unless that's specifically needed.

File: service-api/app/src/App/src/Service/Lpa/FindActorInLpa.php [hunk 4]
```diff
@@ -126,18 +127,35 @@ private function checkForActorMatch(ActorMatchingInterface $actor, array $matchD
...
+            return self::NO_MATCH__DOB;
+        }
+
         $this->logger->debug(
             'Doing actor data comparison against actor with id {actor_id}',
             [
                 'actor_id'   => $actor->getUid(),
                 'to_match'   => $matchData,
-                'actor_data' => array_merge($actorData, ['dob' => $actor->getDob()]),
+                'actor_data' => array_merge($actorData, [
```
Comment (mattmachell) [inline service-api/app/src/App/src/Service/Lpa/FindActorInLpa.php:150]: Please remove this as we don't want PII in logs.

**Review by mattmachell — CHANGES_REQUESTED** [review]

File: service-api/app/src/App/src/Service/Lpa/FindActorInLpa.php (hunk 4 above)
Comment (allenannom) [reply service-api/app/src/App/src/Service/Lpa/FindActorInLpa.php:150]: debug is switched off in prod and this was existing logic that was refactored in the combined LPA tickets, specifically, "getDob()".  The updated test case covers the bug issue now, but not sure if this logic here will still be needed going forward though

**Review by cooperaj — APPROVED** [review]

**Review by mattmachell — DISMISSED** [review]

**Review by cooperaj — APPROVED** [review]
